### PR TITLE
 kvcoord: Restart ranges on a dedicated goroutine.

### PR DESCRIPTION
**Pull request by miretskiy**
Restart ranges on a dedicated goroutine (if needed).
Fix logic bug in stuck range handling.
Increase verbosity of logging to help debug mux rangefeed issues.

Informs #99560
Informs #99640
Informs #99214
Informs #98925
Informs #99092
Informs #99212
Informs #99910
Informs #99560

Release note: None

### CODE REVIEW

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go [hunk 1]
```diff
@@ -554,14 +554,29 @@ func handleRangefeedError(ctx context.Context, err error) (rangefeedErrorInfo, e
...
-	ds.metrics.RangefeedCatchupRanges.Inc(1)
 	catchupSem.SetLimit(maxConcurrentCatchupScans(&ds.st.SV))
-	return catchupSem.Begin(ctx)
+	res, err := catchupSem.Begin(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ds.metrics.RangefeedCatchupRanges.Inc(1)
+	return catchupAlloc(func() {
+		ds.metrics.RangefeedCatchupRanges.Dec(1)
```
Comment (pavelkalinnikov) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:577]: There is another `Dec` for this metric in `singleRangeFeed` func, which is now duplicated. Remove that other `Dec`?

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go [hunk 2]
```diff
@@ -554,14 +554,29 @@ func handleRangefeedError(ctx context.Context, err error) (rangefeedErrorInfo, e
...
+type catchupAlloc func()
+
+// Release implements limit.Reservation
+func (a catchupAlloc) Release() {
+	a()
+}
+
 func acquireCatchupScanQuota(
 	ctx context.Context, ds *DistSender, catchupSem *limit.ConcurrentRequestLimiter,
 ) (limit.Reservation, error) {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:567]: Consider setting the return type to `catchupAlloc` instead of `limit.Reservation`. Also, since it's passed down the stack, replace the type there too.
Technically it's the same interface, so it's compatible, but hiding it behind `limit.Reservation` can be confusing.

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go [hunk 3]
```diff
@@ -554,14 +554,29 @@ func handleRangefeedError(ctx context.Context, err error) (rangefeedErrorInfo, e
 	}
 }
 
+// catchup alloc is a catchup scan allocation.
+type catchupAlloc func()
```
Comment (pavelkalinnikov) [inline pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:558]: general nit: This commit is a mix of 2 fixes - the counter leak and the context. I'd prefer them to be in separate commits or PRs, so that the 2 fixes could be reviewed separately, and (if ever needed) rolled-back/cherry-picked/backported on individual basis.

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go (hunk 2 above)
Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:567]: Sure.

File: pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go (hunk 3 above)
Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_rangefeed.go:558]: Split.

**Comment by miretskiy** [comment]
@pavelkalinnikov all done.

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go [hunk 4]
```diff
@@ -462,10 +462,19 @@ func (m *rangefeedMuxer) restartActiveRangeFeed(
...
-		return divideSpanOnRangeBoundaries(ctx, m.ds, rs, active.startAfter, m.startSingleRangeFeed)
-	}
-	return m.startSingleRangeFeed(ctx, rs, active.startAfter, active.token)
+	// Restart rangefeed on another goroutine.  There are two reasons for doing this
+	// under muxer context group.
+	//   1. Restart might be a bit expensive, particularly if we have to resolve span.
+	//      We do not want to block receiveEventsFromNode for too long.
+	//   2. Most importantly: the context passed to this function may be canceled if
+	//      the restart is happening because node mux is exiting.
+	m.g.GoCtx(func(ctx context.Context) error {
```
Comment (erikgrinaker) [inline pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:471]: Drive-by: isn't there a risk here that we'll spawn a bunch of goroutines whenever a rangefeed fails (see e.g. the `closeWithRestart` call), and isn't that kind of goroutine burst exactly what we're trying to avoid with the mux rangefeeds?

Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:471]: sigh... yeah... And.. I don't even know if this is a problem... like, sure, ctx will get canceled
shortly.  But each of those streams must be restarted against another mux node... Or new connect (with new context) must be established.  I really need to recreate this deadlock, but no such luck so far.

**Comment by miretskiy** [comment]
Moving this to draft.

**Comment by miretskiy** [comment]
@erikgrinaker @pavelkalinnikov Good point regarding spinning up too many Go routines.
Still, I think it's a good idea not to block receive go routine.  Code reworked to make the burst
less likely (and use only 1 go routine when restarting node or stuck rangefeeds).

I still have not been able to reproduce any of those issues.  Part of the reasons is that roachprod-stress
does not work with many issues mentioned here.
I have made a temporary change to log every debug message  in the mux rangefeed.  
I wonder if we truly see a deadlock, or the retry loop is incorrect somehow and keeps retrying even after
test is finished...

**Review by erikgrinaker — APPROVED** [review]
LGTM modulo @pavelkalinnikov's comments.

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go [hunk 5]
```diff
@@ -521,29 +541,19 @@ func (c *muxStream) deleteStream(streamID int64) {
...
 
-func (c *muxStream) closeWithRestart(
-	ctx context.Context, reason error, restartFn func(a *activeMuxRangeFeed) error,
-) error {
+// close closes mux stream returning the list of active range feeds.
+func (c *muxStream) close() []*activeMuxRangeFeed {
 	c.mu.Lock()
 	c.mu.closed = true
-	toRestart := c.mu.streams
+	toRestart := func() []*activeMuxRangeFeed {
```
[the remark this answers is not in the document]
Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:548]: Done.

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go [hunk 6]
```diff
@@ -72,7 +72,7 @@ func muxRangeFeed(
 	catchupSem *limit.ConcurrentRequestLimiter,
 	eventCh chan<- RangeFeedMessage,
 ) (retErr error) {
-	if log.V(1) {
+	if log.V(0) {
```
[the remark this answers is not in the document]
Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:75]: Yes, I intend to roll it back; added todo.

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go [hunk 7]
```diff
@@ -422,24 +434,38 @@ func (m *rangefeedMuxer) receiveEventsFromNode(
 		// get stuck in the first place.
 		if timeutil.Now().Before(nextStuckCheck) {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:435]: Unrelatedly to this PR, which is this not `Now().After(nextStuckCheck)` or `!Now().Before(nextStuckCheck`?

It looks like we're checking the stuck-ness all the time before the `nextStuckCheck` timestamp, and don't check after it. If there is a stuck stream and we miss the `nextStuckCheck` timestamp, we won't restart it. Seems like it should be the other way around?

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go [hunk 8]
```diff
@@ -422,24 +434,38 @@ func (m *rangefeedMuxer) receiveEventsFromNode(
...
-				for _, a := range ms.purgeStuckStreams(threshold) {
-					if err := m.restartActiveRangeFeed(ctx, a, errRestartStuckRange); err != nil {
-						return err
-					}
-				}
+				// Restart rangefeed on another goroutine. Restart might be a bit
+				// expensive, particularly if we have to resolve span.  We do not want
+				// to block receiveEventsFromNode for too long.
+				toRestart := ms.purgeStuckStreams(threshold)
+				m.g.GoCtx(func(ctx context.Context) error {
```
Comment (pavelkalinnikov) [inline pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:441]: Could maybe skip spinning up a goroutine if `len(toRestart) == 0` (which should be the common fast-path?).

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go (hunk 7 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:435]: Also, since `nextStuckCheck` is updated below, we're always checking this. Except when we accidentally miss it - then we won't ever check it again.

Could this cause the deadlocks we're seeing?

Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:435]: Very much related to this PR.... WOW.
I don't know if this is what's causing stuckness... I'm going to fix it now...

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go (hunk 8 above)
Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go:441]: Definitely.

File: pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go (hunk 7 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go]: I don't think we would be restarting them all the time (see `purgeStuckStreams` which seems to be checking the threshold, so most of the time it would return an empty slice), but we would be checking whether the streams are stuck on **each** iteration of this loop. This is wasteful, but it doesn't seem to be incorrect.

However, if we are slow in one of the iterations, and loop around with `Now()` past the `nextStuckCheck`, we would permanently stop checking this, and rangefeeds would never restart. **This** would be problematic.

Comment (miretskiy) [reply pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go]: Ack.

**Review by pav-kv — APPROVED** [review]

**Comment by pavelkalinnikov** [comment]
Please update the description

**Comment by miretskiy** [comment]
> Please update the description

Done.
